text/0000-let-else.md

```diff
@@ -0,0 +1,520 @@
+While `if {} else {}` is technically feasible this RFC proposes it be disallowed for programmer clarity to avoid an `... else {} else {}` situation.
+Rust already provides us with such a restriction, [`ExpressionWithoutBlock`][expressions].
+
+Any pattern that could be put into if-let's pattern position can be put into let-else's pattern position, except a match to a boolean when
+a lazy boolean operator is present (`&&` or `||`), for reasons noted in [future-possibilities][].
+
+If the pattern is irrefutable, rustc will emit the `irrefutable_let_patterns` warning lint, as it does with an irrefutable pattern in an `if let`.
+
+The `else` block must diverge. This could be a keyword which diverges (returns `!`), or a panic.
+This likely necessitates a new subtype of `BlockExpression`, something like `BlockExpressionDiverging`.
+Allowed keywords:
+- `return`
+- `break`
+- `continue`
+
+If the pattern does not match, the expression is not consumed, and so any existing variables from the surrounding scope are
+accessible as they would normally be.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## The diverging block
+
+"Must diverge" is an unusual requirement, which doesn't exist elsewhere in the language as of the time of writing, 
+and might be difficult to explain or lead to confusing errors for programmers new to this feature.
+
+This also necessitates a new block expression subtype, something like `BlockExpressionDiverging`.
+
+## `let PATTERN = if {} else {} else {};`
+
+One unfortunate combination of this feature with regular if-else expressions is the possibility of `let PATTERN = if { a } else { b } else { c };`.
+This is likely to be unclear if anyone writes it, but does not pose a syntactical issue, as `let PATTERN = if y { a } else { b };` should always be
+interpreted as `let Enum(x) = (if y { a } else { b });` (still a compile error as there no diverging block: `error[E0005]: refutable pattern in local binding: ...`)
+because the compiler won't interpret it as `let PATTERN = (if y { a }) else { b };` since `()` is not an enum.
+
+This can be overcome by making a raw if-else in the expression position a compile error and instead requiring that parentheses are inserted to disambiguate:
+`let PATTERN = (if { a } else { b }) else { c };`.
+
+Rust already provides us with such a restriction, and so the expression can be restricted to be a [`ExpressionWithoutBlock`][expressions].
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+let-else attempts to be as consistent as possible to similar existing syntax.
+
+Fundamentally it is treated as a `let` statement, necessitating an assignment and the trailing semicolon.
+
+Pattern matching works identically to if-let, no new "negation" pattern matching rules are introduced.
+
+Operator precedence with `&&` in made to be like if-let, requiring that a case which is an error prior to this RFC be changed to be a slightly different error.
```

Comment:
"in made to be like" doesn't parse and I'm not totally sure what you mean here. Can you specify the current error and new proposed error? This seems likely to be a confusing gotcha, since it's a valid expression but not allowed in `if let`, so we should either find a way to make it explainable, or just allow it and preclude the chaining feature.

Reply:
Currently, `let true = true && false else {};` is a parse error, because that `else` thing doesn't exist. The proposal seems to be  to make this a parse error, which says "`let` expressions in this position are experimental" in the same way as `if let true = true && false {}`. Syntactically, I believe this means that `&&` cannot be used at the top level `EXPR` in `let PAT = EXPR else BLOCK;` unless parentheses are added. This doesn't have to be a grammar modification, since it is possible to post-process the expr to check if it uses `&&` and give the diagnostic. A later RFC might change the grammar to support `&&` properly, but this is just opening syntactic space for such an extension.

Reply:
Yeah, the compiler can check anything, but it complicates things. The guide says "an expression goes here", so if certain expressions are actually restricted, we need a beginner-friendly error message. Also, you can't write a macro that takes an `expr` and reliably generates code using let-else. Overall, I'm just not sure it's worth it to put in ad-hoc confusing grammar restrictions for the sake of a hypothetical feature.

Reply:
There is precedent for a syntactic restriction like this: `if Foo {}.foo() {}` doesn't parse. The purpose of these syntactic restrictions is *specifically* to give a beginner-friendly error message. I don't think it needs to be mentioned in the grammar, but there are more of these than you think.

A macro would not run into these issues, because it uses AST-hygienic construction. If you have a macro `macro_rules! m { ($e:expr) => { if $e {} } }` and then do `m!(Foo {}.foo())` there is no error. The same would be true for let-else.

Reply:
Ooh, that's a very interesting point. If this restriction is enforced at the AST lowering level, then macros can be allowed to generate syntax that can't be written normally. OK, macro objection withdrawn.

I'd still like to see us workshop the wording of these error messages. Is it only a "top-level" boolean operator that would need to be nixed to clear the way for if-let chaining? So, `let true = a && b else { return }` is a no go but `let (true, Some(x)) = (a && b, opt)` is OK?

Reply:
Yes, this is nothing a pair of parentheses can't solve, so it shouldn't be a major blocker for users. Regarding the wording, I suppose "`let` expressions in this position are experimental, see issue [#53667](https://github.com/rust-lang/rust/issues/53667)", borrowed from the error on `if let true = true && false {}`, is a little vague. I would suggest something like

> pattern guards in let-else expressions are [experimental/unsupported]
> see issue [#53667](https://github.com/rust-lang/rust/issues/53667)
> Note: surround the expression in parentheses: `(true && false)`